File: 516.最长回文子序列.py
#
# @lc app=leetcode.cn id=516 lang=python3
#
# [516] 最长回文子序列
#

# @lc code=start
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        dp = [[0]*n for _ in range(n)]
        for i in range(n):
            dp[i][i] = 1
        for l in range(2, n+1):
            for i in range(n-l+1):
                if s[i] == s[i+l-1]:
                    dp[i][i+l-1] = dp[i+1][i+l-2]+2
                else:
                    dp[i][i+l-1] = max(dp[i][i+l-2], dp[i+1][i+l-1])
        return dp[0][n-1]        
# @lc code=end

File: 516.最长回文子序列.cpp
/*
 * @lc app=leetcode.cn id=516 lang=cpp
 *
 * [516] 最长回文子序列
 */

// @lc code=start
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        int n = s.size();
        int dp[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                dp[i][j] = 0;
            }
        }
        for (int i = 0; i < n; i++) dp[i][i] = 1;
        for (int len = 2; len <= n; len++) {
            for (int i = 0; i + len <= n; i++) {
                if (s[i] == s[i+len-1]) dp[i][i+len-1] = 2+dp[i+1][i+len-2];
                else dp[i][i+len-1] = max(dp[i][i+len-2], dp[i+1][i+len-1]);
            }
        }
        return dp[0][n-1];
    }
};
// @lc code=end

